pkg/cloud/aws/actuators/machine_scope.go

```diff
@@ -94,31 +95,33 @@ func (m *MachineScope) Region() string {
 	return m.Scope.Region()
 }
 
-func (m *MachineScope) storeMachineStatus() error {
+func (m *MachineScope) storeMachineStatus(machine *clusterv1.Machine) (*clusterv1.Machine, error) {
 	ext, err := v1alpha1.EncodeMachineStatus(m.MachineStatus)
 	if err != nil {
-		return err
-	}
-
-	m.Machine.Status.ProviderStatus = ext
-
-	if _, err := m.MachineClient.UpdateStatus(m.Machine); err != nil {
-		return err
+		return nil, err
 	}
 
-	return nil
+	machine.Status.ProviderStatus = ext
+	return m.MachineClient.UpdateStatus(machine)
 }
 
 func (m *MachineScope) Close() {
 	if m.MachineClient == nil {
 		return
 	}
 
-	if _, err := m.MachineClient.Update(m.Machine); err != nil {
-		klog.Errorf("[machinescope] failed to update machine: %v", err)
+	latestMachine, err := m.MachineClient.Get(m.Name(), metav1.GetOptions{})
```

Comment:
Shouldn't this be done after the update? It looks like the Update below is actually a noop, and if it somehow did manage to Update the object it would cause the following UpdateStatus to fail.

Reply:
During my tests, it actually showed that the machine somehow gets updated before this method gets called. The `Update` below returns the latest version of the Machine (if successful) with the changes.

After trying and debugging in different stages, the conclusion I came up with is that these calls have to be chained. There is definitely something weird about the object being updated somewhere else during the first run, that's why we need to get the latest version. That said, doing the call even on subsequent calls seems to be fine.